Qnx u mr0 master

yhe39 opened the pull request.

sysopenci commented:
Improper Commit Message
Tracked on not found in commit message,
    make sure Tracked-On: Jira-ticket  is present.

sysopenci commented:
Improper Commit Message
Tracked on not found in commit message,
    make sure Tracked-On: Jira-ticket  is present.

sysopenci commented:
Improper Commit Message
Tracked on not found in commit message,
    make sure Tracked-On: Jira-ticket  is present.

sysopenci commented:
Improper Commit Message
Tracked on not found in commit message,
    make sure Tracked-On: Jira-ticket  is present.

yhe39 committed:
Add the drv_kms_ back

So far in the following two scenario, need use virtio
to allocat scanout buffer, otherwise Android boot with
black screen.
1. On qnx, the virtio is ivshmem, it has blob feature. But
ivshm and i915 memory are in different region, ivshm couldn't
import i915 memory.
2. On redhat, the virtio is virtio-gpu, it do not have blob
on kernel 6.8. The virtio couldn't import buffer from i915
although they are in the same memroy region.

Test Done:
1. Android VM boot without issue.
2. Android display well on QNX.
3. Video playback without issue.

Tracked-On: OAM-119826
Signed-off-by: He, Yue <yue.he@intel.com>

yhe39 committed:
Add the function to check virtio features

If it is virtio-gpu with blob, close drv_kms_.
use i915 to allocate all buffers.
Otherwise use drv_kms_ to allocat scanout non-video
buffers.

Test Done:
1. Android VM boot without issue.
2. Android display well on QNX.
3. Video playback without issue.

Tracked-On: OAM-119826
Signed-off-by: Weifeng Liu <weifeng.liu@intel.com>
Signed-off-by: He, Yue <yue.he@intel.com>

feijiang1 commented on virtgpu_virgl.c:
```diff
@@ -1157,6 +1157,10 @@ static uint32_t virgl_get_max_texture_2d_size(struct driver *drv)
 }
 
+static bool virgl_query_dev_feature() {
+       return (dev_feature == 2);
```
It is easy to introduce problem by just checking the dev_feature counter to judge query_dev_feature enabling

yhe39 replied:
The feature check part:
			if (strcmp(params[i].name, "VIRTGPU_PARAM_QUERY_DEV") == 0) {
				dev_feature += params[i].value;
			}
			if (strcmp(params[i].name, "VIRTGPU_PARAM_RESOURCE_BLOB") == 0) {
				dev_feature += params[i].value;
			}
it doesn't collect feature before, i modify to check the driver name and blob feature.
if has blob feature and is virtio-gpu device, could use i915 to allocate buffer, otherwise need use virtio to allocate scanout buffer.
Do you have better solution?

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -268,6 +268,17 @@ cros_gralloc_driver::cros_gralloc_driver()
 			DRV_INIT(drv_video_, gpu_grp_type, video_idx)
 		if ((virtio_node_idx != -1) && (virtio_node_idx != renderer_idx))
 			DRV_INIT(drv_kms_, gpu_grp_type, virtio_node_idx)
+		if (drv_kms_) {
```
drv_kms_ can i915 node, then it doesn't make sense to call i915 node with drv_query_dev_feature

yhe39 replied:
actually in the drv.c
bool drv_query_dev_feature(struct driver * drv)
{
        bool ret = false;
        assert(drv);
        assert(drv->backend);

        if (drv->backend->query_dev_feature) {
                ret = drv->backend->query_dev_feature();
        }
        return ret;
}
if driver do not have the backend query_dev_feature implemnt. it will return false.
Anyway i will modify the if condition.

feijiang1 reviewed (commented):
need refine

feijiang1 reviewed (commented):
please add more description when use drv_kms_, and why it is needed

yhe39 commented:
> please add more description when use drv_kms_, and why it is needed

added, please check

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -299,6 +317,9 @@ bool cros_gralloc_driver::get_resolved_format_and_use_flags(
 	struct combination *combo;
 
 	struct driver *drv = is_video_format(descriptor) ? drv_video_ : drv_render_;
+	if ((descriptor->use_flags & BO_USE_SCANOUT) && !(IsSupportedYUVFormat(descriptor->droid_format)))
```
switch to check with is_video_format

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -342,6 +363,8 @@ bool cros_gralloc_driver::is_supported(const struct cros_gralloc_buffer_descript
 	uint32_t resolved_format;
 	uint64_t resolved_use_flags;
 	struct driver *drv = is_video_format(descriptor) ? drv_video_ : drv_render_;
+	if ((descriptor->use_flags & BO_USE_SCANOUT) && !(IsSupportedYUVFormat(descriptor->droid_format)))
```
same

feijiang1 commented on virtgpu.c:
```diff
@@ -49,6 +52,14 @@ static int virtgpu_init(struct driver *drv)
+			}
+			if (strcmp(params[i].name, "VIRTGPU_PARAM_RESOURCE_BLOB") == 0) {
+				dev_feature += params[i].value;
+			}
```
suggest use bitwise combination for check, for example:
#define VIRTGPU_PARAM_QUERY_DEV_BIT                (1ull << VIRTGPU_PARAM_QUERY_DEV)
			if (strcmp(params[i].name, "VIRTGPU_PARAM_QUERY_DEV") == 0 && params[i].value == 1) {
				dev_feature |= VIRTGPU_PARAM_QUERY_DEV_BIT;
			}

feijiang1 reviewed (commented):
Need explain more for the difference of virito+ivshemem and virtio-gpu case.

feijiang1 commented on drv.c:
```diff
@@ -833,3 +833,15 @@ uint32_t drv_get_max_texture_2d_size(struct driver *drv)
 	return UINT32_MAX;
 }
+
+bool drv_query_dev_feature(struct driver * drv)
```
the function name seems incorrect, you are checking if blob feature supported in virtio-gpu driver.

feijiang1 reviewed (approved):
LGTM

phreer commented on virtgpu.c:
```diff
@@ -28,11 +28,14 @@ struct virtgpu_param params[] = {
 extern const struct backend virtgpu_virgl;
 extern const struct backend virtgpu_cross_domain;
 
+uint64_t dev_feature = 0;
```
It would be better to make `dev_feature` local to driver specific data. Otherwise, if we have multiple virtio-GPU devices with distinct feature set, this variable can be overwritten.

phreer commented on virtgpu_virgl.c:
```diff
@@ -1157,6 +1158,10 @@ static uint32_t virgl_get_max_texture_2d_size(struct driver *drv)
 		return VIRGL_2D_MAX_TEXTURE_2D_SIZE;
 }
 
+static bool virgl_virtgpu_with_blob() {
```
The name of this function is a little bit confusing as it also requires QUERY_DEV feature. Why does QUERY_DEV matter?

yhe39 replied:
QUERY_DEV  is to check the virtio device name.
you could check the kernel pr: https://github.com/projectceladon/linux-intel-lts2022/pull/36

chenyanxzhu commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -248,15 +250,29 @@ cros_gralloc_driver::cros_gralloc_driver()
 	for (int i = 0; i < availabe_node; i++) {
 		free(node_name[i]);
-		if ((i != renderer_idx) && (i != video_idx)) {
+		if ((i != renderer_idx) && (i != video_idx) && ((drv_kms_ == drv_render_) && (i == virtio_node_idx))) {
```
This line kind of confusing. Shouldn't it be clear like this: 
if ((i != renderer_idx) && (i != video_idx) && (i != virtio_node_idx)) {
and also in this circumstance, the virtio fd already closed in line 258

yhe39 replied:
in the case virtio_node_idx != renderer_idx
when we need use kms to allocate scanout buffer, shouldn't close the virtio_node_index. 
when we do not use kms to allocate scanout buffer, should close the virtio_node_index.
The line 258 you mentioned is the kms driver init fail case.
The condition ensure when the kms is not used to close virtio node.

phreer commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -248,15 +250,40 @@ cros_gralloc_driver::cros_gralloc_driver()
+		if ((virtio_node_idx != -1) && (virtio_node_idx != renderer_idx))
+			DRV_INIT(drv_kms_, gpu_grp_type, virtio_node_idx)
+		if (drv_kms_ && (virtio_node_idx != renderer_idx)) {
+			bool virtiogpu_with_blob = drv_virtgpu_with_blob(drv_kms_);
```
With this change we will allocate scan-out buffers on those virtio-GPU devices that support blob but doesn't support QUERY_DEV, is that what we want?

Looks like you want to distinguish normal virtio-GPU devices from ivshmem-backed devices, but I think `drv_virtgpu_with_blob` is not enough. If I understand correctly, we will be able to query the device name with a new ioctl if QUERY_DEV feature is present. So why don't we call this ioctl and determine whether it's ivishmem device according to its name?

chenyanxzhu commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -268,6 +268,17 @@ cros_gralloc_driver::cros_gralloc_driver()
 			DRV_INIT(drv_kms_, gpu_grp_type, virtio_node_idx)
+		if (drv_kms_ && (virtio_node_idx != renderer_idx)) {
+			bool virtiogpu_with_blob = drv_virtgpu_with_blob(drv_kms_);
+			if (virtiogpu_with_blob) {
```
Maybe add some comment here to clarify. Like: Only virtio device with blob func can share i915 bo to speed up, otherwise will use default virtio way for allocating scanout bo.

yhe39 replied:
they are in the commit message, do you mean we also add comment in code?

chenyanxzhu replied:
Yeah. Just suggestions.

chenyanxzhu commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -454,7 +481,6 @@ int32_t cros_gralloc_driver::allocate(const struct cros_gralloc_buffer_descripto
 		hnd->fds[i] = -1;
 
 	hnd->num_planes = num_planes;
-	hnd->from_kms = false; // not used, just set a default value. keep this member to be backward compatible.
```
from_kms is handle specific. not driver specific. should not make it as driver structue's member

yhe39 replied:
ok

chenyanxzhu commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -546,6 +572,8 @@ int32_t cros_gralloc_driver::retain(buffer_handle_t handle)
 		.use_flags = hnd->use_flags,
 	};
 	drv = is_video_format(&descriptor) ? drv_video_ : drv_render_;
+	if (from_kms)
```
Maybe you can check by: if ((hnd->use_flags & BO_USE_SCANOUT) && (!is_video_format(descriptor)))
and remove "from_kms"

yhe39 replied:
ok

phreer reviewed (approved):
Looks good to me!

feijiang1 reviewed (approved):
LGTM

feijiang1 reviewed (approved):
LGTM

sysopenci commented:
Android CI has started Engineering Build for this issue ,Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
SUCCESS: Android CI has completed Engineering Build for this issue.Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
Android CI has started MERGE Build for this pr ,Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
Android CI has completed MERGE Build for this pr, build is FAILURE. Please check the linked Tracked-On issue/Android CI Web  for more details.
